### PR TITLE
Fix the edge case when y = 0 in kl_div

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #82714

Brought up in https://github.com/pytorch/pytorch/pull/80334#issuecomment-1193600883

We also prepare its opinfo to fix https://github.com/pytorch/pytorch/issues/80488

### CODE REVIEW

**Comment by Lezcano** [comment]
This PR incurs in a regression on CPU (I don't know why), and a small regression on GPU (it looks like we're calling one more kernel.
```
[---- kl_div cpu ---]
               |   PR   | master
4 threads: -------------------
      fwd      |  1100  |  330
      bwd      |  1400  |  280
      fwd+bwd  |  2800  |  720

Times are in milliseconds (ms).

[--- kl_div cuda ---]
               |   PR  | master
4 threads: --------------------
      fwd      |  132  | 132
      bwd      |  376  | 263
      fwd+bwd  |  511  | 395

Times are in microseconds (us).
```

**Comment by Lezcano** [comment]
@jbschlosser PTAL

**Review by jbschlosser — COMMENTED** [review]
Hm, I'm a little concerned about the magnitude of the perf regression, although it is a loss function and it is mostly a CPU thing.

Also, isn't this technically BC-breaking since we now restrict against integral types? Is the reasoning for this that xlogy doesn't support them?

**Comment by Lezcano** [comment]
About integral dtypes, the thing is that `kl_div` is a loss function, and we do not support integral types for those functions. This one in particular is to be used with either logarithmic quantities or numbers between [0,1], so it does not make sense to have it working with integral types.
The idea is the same as that coming from restricting its use with complex numbers.

The perf regression is indeed a bit annoying, but, as you mentioned, this being a loss function, it will almost never be a bottle neck. IMO, correctness is more important here. As we see, in the comment reporting this issue, people do hit this corner case of passing distributions with values of zero and they should get the correct result here.

**Comment by jbschlosser** [comment]
> This one in particular is to be used with either logarithmic quantities or numbers between [0,1], so it does not make sense to have it working with integral types.

What about passing a one-hot target in integral form? I think this would be relatively common, and arguably expected to work

**Comment by Lezcano** [comment]
To be able to compute the KL divergence between two probability measures, P and Q, we need [Q >> P](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence#Definition) (P to be absolutely continuous wrt to Q, or, in other words, Q(x) = 0 -> P(x) = 0). As such, if Q is a point probability measure (a one-hot target), we must have that P == Q. So I don't think that case is too interesting.

**Comment by jbschlosser** [comment]
> As such, if Q is a point probability measure (a one-hot target), we must have that P == Q. So I don't think that case is too interesting.

Sorry, I'm a bit confused. IIUC the purpose of using KL-div as a loss function is to get P to be == Q through iterative training. Why is this not an interesting case?

**Comment by Lezcano** [comment]
The thing is that if `Q` is an integer probability measure, the KL divergence between `P` and `Q` is only defined whenever `P == Q`. If `P != Q` and `Q` is a one-hot probability measure, we have that `KL(P || Q) = inf` so it wouldn't be very useful.

**Comment by Lezcano** [comment]
To see this, note that the formula of the KL divergence is `sum_x p(x) log(p(x) / q(x))`, so if `q(x_0) = 0` and `p(x_0) != 0`, then we get a `log(p(x) / 0) = inf`.

**Comment by Lezcano** [comment]
The thing is that if `Q` is an integer probability measure, the KL divergence between `P` and `Q` is only defined whenever `P == Q`. If `P != Q` and `Q` is a one-hot probability measure, we have that `KL(P || Q) = inf` so it wouldn't be very useful.

**Comment by Lezcano** [comment]
@albanD could you have a look at this one?

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -189,9 +189,12 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
...
 Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
   TORCH_CHECK(!input.is_complex() && !target.is_complex(),
-              "kl_div: Complex inputs not supported.")
+              "kl_div: Complex inputs not supported.");
+  TORCH_CHECK(!at::isIntegralType(input.scalar_type(), /*include_bool*/true) &&
+              !at::isIntegralType(target.scalar_type(), /*include_bool*/true),
+              "kl_div: Integral inputs not supported.");
   auto output = log_target ? at::exp(target) * (target - input)
-                           : target * (at::log(target) - input);
+                           : at::xlogy(target, target) - target * input;
```
Comment (albanD) [inline aten/src/ATen/native/Loss.cpp:197]: Can you give more details on how you measure the regression. From local check, it shouldn't be that bad:
```python
In [6]: a = torch.rand(100)

In [7]: %timeit tmp = a * (a.log() - b)
2.59 µs ± 8.23 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [8]: %timeit tmp = torch.xlogy(a, a) - a * b
2.86 µs ± 10.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

**Review by albanD — COMMENTED** [review]
Sounds good on paper!

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:197]: For small inputs, the kernel is memory bound, so that's why the results are similar. See the benchmarks:
```
[ kl_div cpu (100,) ]
               |  PR | master
32 threads: -------- | ----
      fwd      |  15 |  19
      bwd      |  61 |  40
      fwd+bwd  |  82 |  71

Times are in microseconds (us).

[ kl_div cpu (1000, 1000) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |  200 | 169
      bwd      |  400 | 115
      fwd+bwd  |  600 | 280

Times are in microseconds (us).

[ kl_div cuda (100,) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |   35 |  34
      bwd      |  133 | 100
      fwd+bwd  |  216 | 214

Times are in microseconds (us).

[ kl_div cuda (1000, 1000) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |  132 | 132
      bwd      |  485 | 261
      fwd+bwd  |  619 | 392

Times are in microseconds (us).
```
You see that in CUDA it's still memory bound. The only difference comes from calling 1 more kernel in the backward. Well worth doing this to get correctness IMO.

Here's the script:
```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, device, fn, name):
    # a is in the log-space, i.e. [-inf, 0]
    # b is a probability measure, i.e. [0, 1]
    a = torch.randn(*size, device=device)
    a = -torch.nn.functional.softplus(a)
    a.requires_grad_()

    b = torch.randn(*size, device=device).sigmoid()
    b.requires_grad_()

    z = torch.nn.functional.kl_div(a, b, reduction="batchmean")
    print(name, a.shape, device)

    timer = Timer(
        fn,
        globals={"a": a, "b": b, "z": z},
        label=f"kl_div {device} {size}",
        sub_label=name,
        description="PR",
        num_threads=torch.get_num_threads())

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    fns = [("fwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean')"),
           ("bwd", "z.backward(retain_graph=True)"),
           ("fwd+bwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean').backward()")]
    for device, (name, fn) in product(("cpu", "cuda"), fns):
        yield (100,), device, fn, name
        yield (1000, 1000), device, fn, name


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
```

Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:197]: fwiw, I joined the outputs from master and the PR by hand because I'm lazy.

**Review by albanD — APPROVED** [review]
Ok!
Correctness is more important!
